testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java

```diff
@@ -1346,6 +1347,203 @@ public void interactionsAfterClientStreamCancelAreNoops() throws Exception {
     doPingPong(serverListener);
   }
 
+  // Not all transports support the tracer yet
+  protected boolean haveTransportTracer() {
+    return false;
+  }
+
+  @Test
+  public void transportTracer_streamStarted() throws Exception {
+    server.start(serverListener);
+    client = newClientTransport(server);
+    runIfNotNull(client.start(mock(ManagedClientTransport.Listener.class)));
+    MockServerTransportListener serverTransportListener
+        = serverListener.takeListenerOrFail(TIMEOUT_MS, TimeUnit.MILLISECONDS);
+    if (!haveTransportTracer()) {
+      return;
+    }
+
+    // start first stream
+    long firstTimestamp;
+    {
+      {
+        TransportTracer.Stats before = serverTransportListener.transport.getTransportStats();
+        assertEquals(0, before.streamsStarted);
+        assertEquals(0, before.lastStreamCreatedTimeNanos);
+      }
+      ClientStream clientStream = client.newStream(methodDescriptor, new Metadata(), callOptions);
+      ClientStreamListenerBase clientStreamListener = new ClientStreamListenerBase();
+      clientStream.start(clientStreamListener);
+      StreamCreation serverStreamCreation = serverTransportListener
+          .takeStreamOrFail(TIMEOUT_MS, TimeUnit.MILLISECONDS);
+      {
+        TransportTracer.Stats after = serverTransportListener.transport.getTransportStats();
+        assertEquals(1, after.streamsStarted);
+        firstTimestamp = TimeUnit.NANOSECONDS.toMillis(after.lastStreamCreatedTimeNanos);
+        assertThat(System.currentTimeMillis() - firstTimestamp).isAtMost(50L);
+      }
+
+      ServerStream serverStream = serverStreamCreation.stream;
+      serverStream.close(Status.OK, new Metadata());
+    }
+
+    // start second stream
+    {
```

Comment:
this is a pretty uncommon pattern, its a little jarring to see it.

Reply:
I find this to be a good but underutilized pattern. It is maybe a bit too aggressive in this case though: it's not helping much with `before` and `after`.

Reply:
Changed to use `{}` blocks less aggressively.